### PR TITLE
feat: Add optional `serverSideOnly` to protocol definition

### LINKED ISSUE #621: Exclude a "folder" or some files from being generated to the "_client" side
Hi, first of all congratulations for this amazing server you guys built.

My question is: can I exclude some files or a folder not to be generated on the "_client" module ?

What I want is to exclude all the models that are database entities/tables from being generated to "_client" module and maybe use DTOs.

Basically all database models I want them to be generated only on "_server" part and not in "_client" part, and then DTOs to be generated in both parts.

### PR DESCRIPTION
**Pull request by fischerscode**
Adds the optional key `serverSideOnly` to protocol definition.

Setting this to `true`, prevents the class / enum from being generated at the client side.

```yaml
class: Foo
serverSideOnly: true
fields:
  name: String
```

Fix #621 (Requested ignoring folders, but I think this is a better approach.)

## Pre-launch Checklist

- [x] I read the [Contribute](https://docs.serverpod.dev/contribute) page and followed the process outlined there for submitting PRs.
- [x] This update contains only one single feature or bug fix and nothing else. (If you are submitting multiple fixes, please make multiple PRs.)
- [x] I read and followed the [Dart Style Guide](https://dart.dev/guides/language/effective-dart/style) and formatted the code with [dart format](https://dart.dev/tools/dart-format).
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`), and made sure that the documentation follows the same style as other Serverpod documentation. I checked spelling and grammar.
- [x] I added new tests to check the change I am making.
- [x] All existing and new tests are passing.
- [x] Any breaking changes are documented below.

If you need help, consider asking for advice on the [discussion board](https://github.com/serverpod/serverpod/discussions).

## Breaking changes

*none*

### CODE REVIEW

File: tools/serverpod_cli/bin/generator/class_analyzer.dart [hunk 1]
```diff
@@ -658,4 +668,20 @@ class ClassAnalyzer {
 
     return true;
   }
+
+  bool _validateAndParseServerSideOnly(
+      YamlMap documentContents, CodeAnalysisCollector collector) {
+    var serverSideOnlyNode = documentContents.nodes['serverSideOnly'];
```
Comment (fischerscode) [inline tools/serverpod_cli/bin/generator/class_analyzer.dart:674]: The name is (obviously) open for debate.

Comment (vlidholt) [reply tools/serverpod_cli/bin/generator/class_analyzer.dart:674]: I think `serverOnly` seems a bit more concise.

File: tools/serverpod_cli/bin/generator/class_analyzer.dart [hunk 2]
```diff
@@ -658,4 +668,20 @@ class ClassAnalyzer {
...
+    if (serverSideOnlyNode != null) {
+      if (serverSideOnlyNode.value is! bool) {
+        collector.addError(SourceSpanException(
+          'The "serverSideOnly" property must be a Boolean.',
```
Comment (vlidholt) [inline tools/serverpod_cli/bin/generator/class_analyzer.dart:678]: "must be a bool", to avoid any confusion.

**Review by vlidholt — CHANGES_REQUESTED** [review]
This looks great, but it will need a test (with the value both set to true or false).

File: tools/serverpod_cli/bin/generator/class_analyzer.dart (hunk 2 above)
Comment (fischerscode) [reply tools/serverpod_cli/bin/generator/class_analyzer.dart]: https://github.com/serverpod/serverpod/pull/625/commits/0fd38dfcf54bfd025936c2ab0baabc24ea1c96fb

File: tools/serverpod_cli/bin/generator/class_analyzer.dart (hunk 1 above)
Comment (fischerscode) [reply tools/serverpod_cli/bin/generator/class_analyzer.dart]: https://github.com/serverpod/serverpod/pull/625/commits/c7d86bb282299e11024d72023c2ee83abe6d66c0

**Comment by fischerscode** [comment]
@vlidholt I added tests. https://github.com/serverpod/serverpod/pull/625/commits/fdb984f1095556f1b9050dd91baa4c1b479aa9e0 (`generated_test.dart` at client AND server) But currently they are not executed. I'm  not sure where you want them to be triggered. I'd suggest adding them to `run_tests_single_server` using a `dart test generated_test.dart` at the local machine (not in docker). But somehow this doesn't seem right.

File: tests/serverpod_test_server/test/generated_test.dart [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+    });
+    test('Ensure ServerOnlyEnum is available', () async {
+      await expectLater(library.withCode('''
+import 'package:serverpod_test_server/src/generated/protocol.dart';
+
+late ServerOnlyEnum test;
+'''), compiles);
+    });
+  });
+}
```
Comment (fischerscode) [inline tests/serverpod_test_server/test/generated_test.dart:52]: I used so many tests, due to [this](https://github.com/rrousselGit/expect_error/issues/8).

**Review by fischerscode — COMMENTED** [review]
The generated Enums (test) are in the wrong directory (#628)... Blocked by this PR might be considered blocked by #630.

**Comment by vlidholt** [comment]
@fischerscode I think a test could somewhat easily be added here: https://github.com/serverpod/serverpod/blob/main/tests/docker/tests_cli/run-tests.sh

There is a test that runs the tooling and checks that classes are created correctly (or at least that they exist). Just want to make sure that the correct file is added in the server, but not on the client.

**Comment by fischerscode** [comment]
I added the tests as dart tests in fdb984f1095556f1b9050dd91baa4c1b479aa9e0.

Should I revert fdb984f1095556f1b9050dd91baa4c1b479aa9e0 and use `test -f` instead? It might be easier to read.

**Comment by fischerscode** [comment]
The failing analyze should be related to the new version of the [postgres package](https://pub.dev/packages/postgres).

**Comment by fischerscode** [comment]
@vlidholt I've reworked the tests in f6388a2 and cb1f231. I hope the position is right.

**Comment by fischerscode** [comment]
@vlidholt I think 97d4d610d5062516745d3c8f264724624d41df5c would be better then 280833befa444eac8f0c88d01e4480810b867a52. (Missed the templates pubspec)

**Comment by fischerscode** [comment]
Soft blocked by #638...

**Comment by fischerscode** [comment]
This PR now includes 02afa1b. I'm not sure, why the change occurred while regenerating...

**Review by vlidholt — APPROVED** [review]
This looks great! Thank you!

**Pull request merged**